Proper handling of <BR> tags in .html documents

poire-z opened the pull request:
`<BR>` were always being converted to `<P>` because of "Patch for bad LIB.RU books" that was being applied everytime.
So, it was losing the style of the container for text after a `<br>`
With this HTML:
```html
<html>
<head>
<style>
p { color: red; text-indent: 1.2em;}
</style>
</head>
<body>

This is some text.<br/>
This is some text.<br/>
This is some text.<br/>
This is some text.<br/>This is some text.<br/>This is some text.<br/>This is some text.<br/>
This is some text.

<hr/>

<p class=thisisaP style="color: blue">
This is some text.<br/>
This is some text.<br/>
This is some text.
</p>

<hr/>

<div class=thisisaDIV style="color: green">
This is some text.<br/>
This is some text.<br/>
This is some text.
</div>

<hr/>

<div class=thisisaDIV style="color: green">
This is some text.<br/>
This is some text.<br/>
This is some text.
</div>

<hr/>

<pre class=thisisaPRE style="color: teal">


          This is some text.<br/>
This is some text.<br/>

</pre>


</body>
</html>
```


<kbd>![before](https://user-images.githubusercontent.com/24273478/39563823-f9bd220a-4eb1-11e8-97f0-9080381e661f.png)</kbd> => <kbd>![after](https://user-images.githubusercontent.com/24273478/39563827-fcc47dfe-4eb1-11e8-8415-0bffa3cf1cc9.png)</kbd>

There is still a problem with leading space/newline not being stripped out, but I can't figure out (yet) where to fix that (lvtextfm.cpp probably).
So, may be don't merge this yet... watching the screenshots I just made, the wrong behaviour seems less intriguing that the fixed one :| 

Also fix unrelated clang warning (the only serious one I found in travis output), in code from https://github.com/koreader/crengine/commit/d0854294e
```
src/lvimg.cpp:1198:40: warning: & has lower precedence than !=; != will be evaluated first [-Wparentheses]
                if (p[1]==0xf9 && (p[3]&1!=0))
                                       ^~~~~
src/lvimg.cpp:1198:40: note: place parentheses around the '!=' expression to silence this warning
                if (p[1]==0xf9 && (p[3]&1!=0))
                                       ^
                                        (   )
src/lvimg.cpp:1198:40: note: place parentheses around the & expression to evaluate it first
                if (p[1]==0xf9 && (p[3]&1!=0))
                                       ^
                                   (     )
```

poire-z commented:
Mhhhh, looks like it is fixed if I make `<BR>` be declated just like `<HR>` as supposed to be a block element instead of inline (I guess for inline, it keeps space/newline so words are not stick together):

```diff
--- a/crengine/include/fb2def.h
+++ b/crengine/include/fb2def.h
 XS_TAG1I( em )
 XS_TAG1I( q )
 XS_TAG1I( span )
-XS_TAG1I( br )
+XS_TAG1( br )

 XS_TAG1D( title, true, css_d_block, css_ws_normal )

 XS_TAG1I( b )
 XS_TAG1I( i )

 // type="styleType"
 XS_TAG1I( strikethrough )
 XS_TAG1I( sub )
 XS_TAG1I( sup )
 XS_TAG1I( style )
 XS_TAG1I( strong )
 XS_TAG1I( emphasis )
 XS_TAG1D( code, true, css_d_inline, css_ws_pre )
 XS_TAG1I( a )

 XS_TAG1( html )
 XS_TAG1( head )

 XS_TAG1( hr )
```
<kbd>![image](https://user-images.githubusercontent.com/24273478/39564788-e0103e10-4eb5-11e8-9690-e9947cbbaff1.png)</kbd>

But changing this feels a bit scary (and text selection feels a bit off - which it already did previously on some line when BR were involved, but with that, there's more lines involved :)....)

poire-z commented:
About this change:
```
-XS_TAG1I( br )
+XS_TAG1( br )
```
I guess it never mattered if BR was inline or a block element, because for 10 years (since this LIB.RU patch), `<BR>` had been eradicated from crengine :)
So, know that we want to preserve them, and they need to be block elements to work as expected, we need this change.
Do you agree it's logical for `<BR>` to be managed like `<HR>` are ? After all, a `<BR>` is a bit like an invisible `<HR>` with no line and no height :)

There is still a thing that `<br><br><br>` don't add up (they didn't before either), and they display as a single `<br>`, which I don't know if it is the correct behaviour or just some convenience that browsers allow.

Then the other bug that showed up while testing (a leading space, thus being exlucded for text rendering, is still accounted in text selection boxes display) is unrelated to this change.

poire-z commented:
The multiple `<br>` effect could be achieved by adding to our css:
```
br + br {
   margin-top: 0.5em;
   margin-bottom: 0.5em
   }
```
(with whatever proper sizing these should have :)

Frenzie commented:
http://w3c.github.io/html/textlevel-semantics.html#the-br-element

>  While line breaks are usually represented in visual media by physically moving subsequent text to a new line, a style sheet or user agent would be equally justified in causing line breaks to be rendered in a different manner, for instance as green dots, or as extra spacing. 

But normally I'd say a BR should just be a line break.

> Do you agree it's logical for <BR> to be managed like <HR> are ? After all, a <BR> is a bit like an invisible <HR> with no line and no height :)

Well yeah, in behavior it's more or less an invisible block. Empty `HR`, empty `P`, empty `DIV`, it's all fine really as long as it's not affected by styles targeting one of those but only by styles targeting `BR`.

poire-z commented:
OK, rebased and added BR+BR to all css files.

**There's a thing with this: it may break bookmarks and highlights made on older versions :(**

With this , the DOM may change, and stuff that was at:
`/body/DocFragment/body/div/p[10]/text()[3].3`
may now be at:
`/body/DocFragment/body/div/p[9]/autoBoxing/text()[4].3`
so old ones may be invalid or point to other things... (or maybe crash in some cases).

Frenzie commented:
Hm, that's unfortunate. :-/

Crash even? Wouldn't it just be an invalid DocFragment ID?

poire-z commented:
This should allow requesting old behaviour with:
`self._document:setIntProperty("crengine.dom.version", 20171225)
`and new the one with
`self._document:setIntProperty("crengine.dom.version", 20180503)`.

Still the less fun frontend side to do :(

poire-z commented:
Fixed incoherent dates i wrote in comments...:
```
// 20100101 to 20180302: historical version
//
// 20180325: fixed <BR> that were previously converted to <P> because
```

poire-z commented:
I just realized that only non-EPUB (ie: plain HTML) were affected by this BR => P conversion.
EPUBs use a different XML/DOM parser/builder, with strict XML requirements,, that had no LIB.RU tweaks.

The internal DOM is now the same whether I put the same HTML in an EPUB or view it standalone.

So, it was a bit less critical to have this version check/requesting, as I believe HTML documents are less often used (but I don't know, may be people do, and do some highlights in these).
So, let's keep the whole cre_dom_version stuff anyway, just in case we later fix another thing impacting the DOM?

poire-z commented:
Reworded 1st commit to mention it applied only to HTML documents.
OK for merging? (I'll do the full base bump and updates the PR and commit texts this evening)

poire-z commented:
Looks like we still have an issue with `<BR>`...
For this snippet:
```html
<style>P { margin: 0; }</style>
<H1>text BR block</H1>
<div>
<p>abc</p>
<p><br/><br/></p>
<p>def</p>
<hr/>
<p>abc</p>
<p><br/></p>
<p>def</p>
<hr/>
<p>abc</p>
<p><br style="display: block"/><br style="display: block"/></p>
<p>def</p>
<hr/>
<p>abc</p>
<p><br style="display: block"/></p>
<p>def</p>
<hr/>
```

From left to right, rendering by Calibre | Firefox | KOReader:
<kbd>![image](https://user-images.githubusercontent.com/24273478/79199597-82243980-7e35-11ea-9c35-6dfca1e79ddb.png)</kbd>

So, we miss one empty line for the first normal BR, which looks like a bug.

But for BRs with `display:block`, we do just as Firefox (no empty line, no height I guess) - but Calibre renders them (the current book I'm reading has that kind of paragraph separator, and I guess it - and me the reader - expect them to show as empty line.

Any idea what/why/how the differences between calibre & Firefox ?

Frenzie commented:
calibre uses QtWebKit (or maybe now QtWebEngine = Blink), which seems to behave the same. Here's a screenshot from Vivaldi. I'm not really sure what's going on there but also I don't have Firefox on Windows atm.

![vivaldi_873DAopiKM](https://user-images.githubusercontent.com/202757/79211853-0b903780-7e47-11ea-832a-c234e77207c2.png)

poire-z commented:
I'm inclined to think Firefox and KOReader are doing the right thing: if a BR is `display:block`, it's a block. If it has no height specified: it's a 0-height block, and should use no height and make no spacing.

But I think we should render as Calibre/Blink.

I see 2 options for that:

A. force BR to always be `display:inline`, even if the publisher has set it to `display:block` : it will be rendered correctly in a "final" paragraph, with a line height depending on the CSS/inherited font size and line-height.

B. let it be block, but if it has no height specified, give it the height of the font (which is larger than 1em) (+ inherited line-height) : might introduce some font metric computations in the renderBlockElement code where there was none.
We already have that kind of handling for `<EMPTY-LINE>` elements:
https://github.com/koreader/crengine/blob/d623aa6042d6e8862faab50f18bffa6ffc842a04/cr3gui/data/epub.css#L155-L158
https://github.com/koreader/crengine/blob/d623aa6042d6e8862faab50f18bffa6ffc842a04/crengine/src/lvrend.cpp#L6053-L6075

I guess Calibre/Blink took one of these 2 options - but I dunno which.
Any feeling of which option is best / more natural ?

Frenzie commented:
Option B sounds more natural/expected to me but if option A is a lot simpler it also sounds close enough.

Incidentally, here's classic Opera (with the Presto engine).

![Screenshot_2020-04-15_11-56-52](https://user-images.githubusercontent.com/202757/79324673-3befd880-7f10-11ea-983a-a81965b89ccd.png)

poire-z commented:
OK, I'll see about doing B.

> So, we miss one empty line for the first normal BR, which looks like a bug.

Standalone BR with `display:inline` not making an empty line should be fixed with:
```diff
--- a/crengine/src/lvtextfm.cpp
+++ b/crengine/src/lvtextfm.cpp
@@ -1089,9 +1089,12 @@ public:
                     lChar16 c = m_text[pos];

                     bool is_space = (c == ' ');
-                    if ( is_space && prev_was_space && !preformatted ) {
+                    if ( is_space && prev_was_space && !preformatted && src->object ) {
                         // On non-pre paragraphs, flag spaces following a space
                         // so we can discard them later.
+                        // (But only if the space is from a document text node (it then
+                        // has a non-NULL ->object), to keep those we added for empty
+                        // lines or identation with 'txform->AddSourceLine(L" "...)'.)
                         m_flags[pos] = LCHAR_IS_COLLAPSED_SPACE | LCHAR_ALLOW_WRAP_AFTER;
```

NiLuJe commented:
@poire-z: Possible repo in the wild: https://www.mobileread.com/forums/showthread.php?t=328961 ;).

poire-z commented:
Yep, quite likely.
A bit different than my test case above, but same effect, whether the (non wrapped in a P) BR is `display:inline` or `display:block`:
```
<hr/>
<p>abc</p>
<br/>
<p>def</p>
<hr/>
<p>abc</p>
<br style="display: block"/>
<p>def</p>
<hr/>
```
With a month old KOReader (and probably current):
<kbd>![image](https://user-images.githubusercontent.com/24273478/79432917-c7c73a80-7fcc-11ea-969c-90451a1c1617.png)</kbd>
With my not yet commited fixes:
<kbd>![image](https://user-images.githubusercontent.com/24273478/79433023-ea595380-7fcc-11ea-95c4-ef788ce4879e.png)</kbd>

I'll PR that (with the typography stuff) after coming up release.

In the meantime, if calibre1 makes it block, it can be solved with a user style tweak:
`BR { height: 1em; }`
and if not, may be with this one:
`DIV > BR { display: block; height: 1em; }`

poire-z commented:
Still some issues with BR... See https://github.com/koreader/koreader/pull/6069#issuecomment-615914802

With:
```html
<hr/>
<hr/>
<hr/>
<p>abc
<br style="display: block; height: 5em"/>
def</p>
<hr/>
<p>abc
<br style="display: block; height: 5em"></br>
def</p>
<hr/>
<p>abc</p>
<br style="display: block; height: 5em"/>
<p>def</p>
<hr/>
<p>abc</p>
<br style="display: block; height: 5em"></br>
<p>def</p>
<hr/>
<p>real block
<span style="display: block; height: 5em"></span>
span</p>
```

Calibre | Firefox | KOReader(with BR forced to inline) | KOReader with BR being rendered block
(look below the 3 HRs)

<kbd>![image](https://user-images.githubusercontent.com/24273478/79669508-47ffc280-81bc-11ea-91ad-add95690333f.png)</kbd>

So, last on the right is with https://github.com/koreader/crengine/commit/36ae6ccd02f7b0ff8c3c6ac7c93f6c97afadb02a which makes all the BR be 5em and as tall as the last SPAN block test.

So, I guess neither Calibre nor Firefox really handle block BR as really block elements (or they would ensure their CSS height).
So solution B) above must not be the right one.

So, I guess I'll go with resetting them to css_d_inline. (I'll keep the added code for block br, just in case, even if it would never be triggered...)

Still have to figure out why the difference in Calibre/Firefox vs us with:
`<br style="display: block; height: 5em"></br>`
which might be ill-formed - but both shows kinda 2 BRs - while we handle it just like if it were self closing.
I guess I can discard this ill-case, right ?

Frenzie commented:
I think/hope that it's indeed a very niche edge case extremely unlikely to show up in the wild.
